sample/src/main/java/com/polidea/rxandroidble2/sample/example4_characteristic/advanced/Presenter.java

```diff
@@ -0,0 +1,230 @@
+
+
+import android.bluetooth.BluetoothGattCharacteristic;
+import android.support.annotation.NonNull;
+import android.support.v4.util.Pair;
+
+import com.polidea.rxandroidble2.NotificationSetupMode;
+import com.polidea.rxandroidble2.RxBleConnection;
+import com.polidea.rxandroidble2.RxBleDevice;
+
+import java.util.UUID;
+
+import io.reactivex.Observable;
+import io.reactivex.ObservableTransformer;
+import io.reactivex.Single;
+
+import static android.bluetooth.BluetoothGattCharacteristic.PROPERTY_INDICATE;
+import static android.bluetooth.BluetoothGattCharacteristic.PROPERTY_NOTIFY;
+import static java.util.Arrays.asList;
+
+/**
+ * Presenter class for {@link AdvancedCharacteristicOperationExampleActivity}. Prepares the logic for the activity using passed
+ * {@link Observable}s. Contains only static methods to show a purely reactive (stateless) approach.
+ * <p>
+ * The contract of the class is that it subscribes to the passed Observables only if a specific functionality is possible to use.
+ */
+final class Presenter {
+
+    private static UUID clientCharacteristicConfigDescriptorUuid = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb");
+
+    private Presenter() {
+        // not instantiable
+    }
+
+    static Observable<PresenterEvent> prepareActivityLogic(
+            final RxBleDevice rxBleDevice,
+            final UUID characteristicUuid,
+            final Observable<Boolean> connectClicks,
+            final Observable<Boolean> connectingClicks, // used to disconnect the device even before connection is established
+            final Observable<Boolean> disconnectClicks,
+            final Observable<Boolean> readClicks,
+            final Observable<byte[]> writeClicks,
+            final Observable<Boolean> enableNotifyClicks,
+            // used to disable notifications before they were enabled (but after enable click)
+            final Observable<Boolean> enablingNotifyClicks,
+            final Observable<Boolean> disableNotifyClicks,
+            final Observable<Boolean> enableIndicateClicks,
+            // used to disable indications before they were enabled (but after enable click)
+            final Observable<Boolean> enablingIndicateClicks,
+            final Observable<Boolean> disableIndicateClicks
+    ) {
+        return connectClicks.take(1) // subscribe to connectClicks and take one (unsubscribe after)
+                .flatMap(clickEvent ->
+                                rxBleDevice.establishConnection(false) // on click start connecting
+                                        .flatMapSingle(rxBleConnection ->
+                                                getCharacteristic(characteristicUuid, rxBleConnection)
+                                                        .map(bluetoothGattCharacteristic -> Pair.create(rxBleConnection, bluetoothGattCharacteristic)))
+                        // TODO: [PU] 27.02.2018 Darek, do you have any ideas on how to do it better to keep your example's ideas?
```

Comment:
Better in terms of? :D

We could use two `.flatMap`s here:
```
.flatMap(clickEvent -> rxBleDevice.establishConnection(false))
.flatMap(
  connection -> getCharacteristic(characteristicUuid, rxBleConnection),
  connection, characteristic -> Pair.create(connection, characteristic)
)
...
```

If you would like to rearrange the code to make it easier to read / comprehand—feel free to try :)

Reply:
I don't want to, be honest :) getCharacteristic is a Single now and this won't work I guess.

Reply:
so it would need here to have `.toObservable()` I suppose